rbac/management/permission_scope.py

```diff
@@ -0,0 +1,296 @@
+
+        The argument shall be a V1 permission string (as if for _PermissionDescriptor.parse_v1).
+        The permission may be a wildcard.
+
+        Matching precedence (highest to lowest):
+        1. Exact app:resource_type:verb match.
+        2. Wildcard app:resource_type:* match.
+        3. Wildcard app:*:verb match.
+        4. Wildcard app:*:* match.
+        5. Finally, if no match exists, the DEFAULT scope.
+
+        Note that, if the permission is a wildcard, some of these steps will be redundant.
+        For instance, if the permission is app:*:verb, there are only two possible matches:
+        app:*:verb and app:*:*.
+        """
+        parsed = _PermissionDescriptor.parse_v1(permission)
+
+        # If we are passed a wildcard, some wildcard checks will be redundant.
+        candidates = [parsed] + [
+            wildcard
+            for wildcard in [
+                parsed.with_unconstrained_verb(),
+                parsed.with_unconstrained_resource(),
+                parsed.with_app_only(),
+            ]
+            if wildcard != parsed
+        ]
+
+        for candidate in candidates:
+            scope = self._permissions_map.get(candidate)
+
+            if scope is not None:
+                return scope
+
+        return Scope.DEFAULT
+
+    def highest_scope_for_permissions(self, permissions: Iterable[str]) -> Scope:
+        """
+        Return the highest scope to which any permission in permissions is assigned.
+
+        Permission scopes are determined as if by using scope_for_permission.
+        """
+        return max(
+            [self.scope_for_permission(permission) for permission in permissions],
+            default=Scope.DEFAULT,
+        )
+
+    def v2_bound_resource_for_permission(
+        self,
+        permissions: Iterable[str],
+        tenant_org_id: str,
+        root_workspace_id: str,
+        default_workspace_id: str,
+    ) -> V2boundresource:
+        """
+        Return a V2boundresource corresponding the highest scope for any permission in permissions.
+
+        The appropriate scope is determined as if by highest_scope_for_permissions. A
+        V2boundresource is then returned, bound to the appropriate provided resource.
+        """
+        scope = self.highest_scope_for_permissions(permissions)
+
+        if scope == Scope.TENANT:
+            tenant_resource_id = f"{settings.PRINCIPAL_USER_DOMAIN}/{tenant_org_id}"
+            return V2boundresource(resource_type=("rbac", "tenant"), resource_id=tenant_resource_id)
+        elif scope == Scope.ROOT:
+            return V2boundresource(resource_type=("rbac", "workspace"), resource_id=root_workspace_id)
+        elif scope == Scope.DEFAULT:
+            return V2boundresource(resource_type=("rbac", "workspace"), resource_id=default_workspace_id)
+        else:
+            raise AssertionError(f"Unexpected scope: {scope}")
+
+
+_default_service = threading.local()
```

Comment:
why do we need threading.local() here? Can't we just instantiate a _default_service = ImplicitResourceService()?